File: web/src/api/newproject.ts
// Typed data-access for /api/v1/newproject (gm-root.17.3 — see
// docs/design/newproject.md).
//
// The /new SPA route runs a conversational project-creation flow
// against the `newproject` skill. The skill (gm-root.17.5) and the
// Onboarder persona host (gm-root.17.10) do NOT exist yet — this
// module is a typed STUB client. Until the backend lands the e2e
// suite intercepts /api/v1/newproject/** in
// testing/e2e/fixtures/server.ts and the SPA route renders + commits
// against the canned fake.
//
// Contract (per design doc):
//
//   POST /api/v1/newproject/start         → mint session id
//   POST /api/v1/newproject/:id/turn      → submit operator message,
//                                            return updated state +
//                                            skill reply
//   POST /api/v1/newproject/:id/ratify    → nonce-confirmed atomic
//                                            commit
//
// One-shot persistence: state lives only in server memory. Refresh /
// disconnect / restart loses the session. The SPA does NOT autosave
// or attempt to resume — by design.

import { apiFetch } from './client';
import { CONFIRM_HEADER } from './workItems';

// ── Output schema (mirrors the Go side in docs/design/newproject.md) ──
//
// The skill emits a fully-populated plan tree. Every field is
// populated; empty strings, empty slices, and Estimate=0 are valid
// empty states. The "no missing keys" contract makes ratification
// total.

export interface DraftBead {
  Title: string;
  Description: string;
  // "task" | "bug" | "feature" | "chore" — open string at the wire so
  // future bead kinds don't break the SPA.
  Type: string;
  Acceptance: string;
  Labels: string[];
  Priority: number;
  Estimate: number;
  Skills: string[];
  DesignNotes: string;
  Notes: string;
  // Intra-tree refs (e.g. "milestone:0/epic:1/bead:2"). Resolved to
  // real bd-… ids during ratification.
  DependsOnRefs: string[];
  BlocksRefs: string[];
}

export interface DraftEpic {
  Title: string;
  Description: string;
  Acceptance: string;
  Labels: string[];
  Priority: number;
  Estimate: number;
  Skills: string[];
  DesignNotes: string;
  Notes: string;
  Beads: DraftBead[];
}

export interface DraftMilestone {
  Title: string;
  Description: string;
  Acceptance: string;
  Labels: string[];
  Priority: number;
  Estimate: number;
  Skills: string[];
  DesignNotes: string;
  Notes: string;
  Epics: DraftEpic[];
}

// ChangeRef points into the plan tree for the most-recent skill edit
// (used by the diff surfacing in the conversation pane). Open shape —
// the skill emits whatever granularity it needs and the SPA renders
// the path verbatim.
export interface ChangeRef {
  // path is a slash-joined path like "milestone:1/epic:0/bead:2".
  // Empty when the change spans the whole tree (e.g. a fresh draft).
  path: string;
  // kind labels what changed (added | removed | renamed | edited).
  // The SPA tone-codes the diff badge by kind.
  kind: 'added' | 'removed' | 'renamed' | 'edited' | '';
  // summary is a human-readable one-liner ("Renamed milestone 2 to
  // 'OSS-ready'"). Empty when the skill didn't synthesize one.
  summary: string;
}

// NewProjectState is the authoritative server-side state for the
// session. Each turn mutates the state in place; the Ratify button
// commits whatever the state is at click time.
export interface NewProjectState {
  ProjectName: string;
  Description: string;
  TechStack: string[];
  // Free-form architecture notes captured from the operator.
  Architecture: string;
  Milestones: DraftMilestone[];
  // DraftProjectMD is the running synthesis the skill keeps current
  // each turn. Ratification persists this verbatim as docs/project.md.
  DraftProjectMD: string;
  Turn: number;
  LastChange: ChangeRef;
}

// EMPTY_STATE is the initial state the SPA renders before the first
// turn lands. It mirrors the "no missing keys" contract — every field
// is present, just empty.
export const EMPTY_STATE: NewProjectState = {
  ProjectName: '',
  Description: '',
  TechStack: [],
  Architecture: '',
  Milestones: [],
  DraftProjectMD: '',
  Turn: 0,
  LastChange: { path: '', kind: '', summary: '' },
};

// ── Conversation transcript ─────────────────────────────────────────

// ConversationMessage is one turn in the message history. The skill
// emits 'assistant' messages (with optional diff metadata in the
// state's LastChange); the operator emits 'user' messages.
export interface ConversationMessage {
  // Stable id so React lists key cleanly even when the same skill
  // reply is re-rendered between turns.
  id: string;
  role: 'user' | 'assistant';
  // Display content. The SPA renders this as plain text in v1; future
  // iterations may upgrade to markdown once the skill emits structured
  // bullets.
  content: string;
  // ISO8601 — server-supplied for assistant messages, client-supplied
  // for user messages.
  at: string;
}

// ── Wire envelopes ──────────────────────────────────────────────────

export interface StartNewProjectResponse {
  // Opaque session id used to scope subsequent /turn and /ratify
  // calls. Lives only in server memory.
  session_id: string;
  // Initial state — empty per EMPTY_STATE on day 0, but the wire
  // shape leaves room for the skill to seed defaults later.
  state: NewProjectState;
  // Optional greeting from the skill. Empty string on day 0.
  greeting: string;
}

export interface TurnRequest {
  // Operator message. Free text — the skill parses intent.
  message: string;
  // Optional in-place edits the operator made to the plan preview
  // pane. The skill sees these as "the operator already changed X" so
  // its next turn doesn't try to re-derive them. Open shape — the
  // skill interprets the path-keyed diff.
  edits?: Record<string, unknown>;
}

export interface TurnResponse {
  // Updated full state. The wire is authoritative — the SPA replaces
  // its local copy each turn rather than merging diffs.
  state: NewProjectState;
  // Skill reply for the conversation pane. Empty when the skill
  // chooses to update the plan silently (rare).
  reply: string;
  // Server-assigned message id for the reply (so the SPA's transcript
  // can dedupe + key without re-deriving).
  reply_id: string;
  // Server-supplied timestamp for the reply.
  reply_at: string;
}

export interface RatifyRequest {
  // Snapshot of the state the operator confirmed in the modal. The
  // server compares this against the live session state and rejects
  // mismatches so a stale modal can't commit a state the operator
  // didn't see.
  state: NewProjectState;
}

export interface RatifyResponse {
  // Filesystem path of the new project root
  // (default_dir/<project-name>/).
  project_path: string;
  // Human-readable project name — echoed back by the server so the
  // handoff screen can confirm which project was created without
  // reading it from local state.
  project_name: string;
  // Milestone and epic counts — seeded by the ratification transaction
  // so the handoff screen can show "seeded N milestones and M epics"
  // without re-querying the new workspace's beads database. Both 0
  // for the lightweight create path that bypasses the conversational
  // flow (gm-root.17.13).
  milestone_count: number;
  epic_count: number;
}

// gm-root.17.13 — lightweight project creation. The body is two
// fields: a project name and an optional description. The server
// scaffolds the dir + git + .gemba + beads DB with no plan tree
// (empty Milestones[]). No Onboarder, no LLM. The same
// X-GEMBA-Confirm nonce gate the conversational /ratify uses
// applies; the response shape is the same RatifyResponse so the
// handoff screen treats both paths uniformly.
export interface CreateProjectRequest {
  project_name: string;
  description: string;
}

// gm-root.17.13 — Onboarder availability probe. The board's empty
// state reads this to decide whether to render "Plan with the
// Onboarder →". 200 in both available + unavailable cases; the
// `available` flag is the operative signal.
export interface OnboarderProbeResponse {
  available: boolean;
  reason?: string;
}

// ── Client functions ────────────────────────────────────────────────

// startNewProject — POST /api/v1/newproject/start. Mints a fresh
// session id and returns the initial state. No request body in v1
// (the skill kicks off the conversation with a generic greeting); a
// future iteration may pass a seed prompt.
export async function startNewProject(): Promise<StartNewProjectResponse> {
  return apiFetch<StartNewProjectResponse>('/v1/newproject/start', {
    method: 'POST',
    body: '{}',
    headers: { 'Content-Type': 'application/json' },
  });
}

// submitTurn — POST /api/v1/newproject/:id/turn. Submits an operator
// message (and any in-place plan edits) and returns the updated state
// plus the skill reply.
export async function submitTurn(
  sessionId: string,
  body: TurnRequest
): Promise<TurnResponse> {
  if (!sessionId) throw new Error('submitTurn: sessionId is required');
  return apiFetch<TurnResponse>(`/v1/newproject/${encodeURIComponent(sessionId)}/turn`, {
    method: 'POST',
    body: JSON.stringify(body),
    headers: { 'Content-Type': 'application/json' },
  });
}

// ratifyNewProject — POST /api/v1/newproject/:id/ratify. Nonce-gated
// through the shared X-GEMBA-Confirm header so a double-click on the
// confirm button can't double-write the project. The server runs the
// atomic transaction described in the design doc and returns the SPA's
// post-ratify destination.
export async function ratifyNewProject(
  sessionId: string,
  body: RatifyRequest,
  opts: { nonce?: string } = {}
): Promise<RatifyResponse> {
  if (!sessionId) throw new Error('ratifyNewProject: sessionId is required');
  return apiFetch<RatifyResponse>(`/v1/newproject/${encodeURIComponent(sessionId)}/ratify`, {
    method: 'POST',
    body: JSON.stringify(body),
    headers: {
      'Content-Type': 'application/json',
      [CONFIRM_HEADER]: opts.nonce ?? freshNonce(),
    },
  });
}

export function freshNonce(): string {
  if (typeof crypto !== 'undefined' && typeof crypto.randomUUID === 'function') {
    return crypto.randomUUID();
  }
  return `nonce-${Date.now()}-${Math.random().toString(36).slice(2, 10)}`;
}

// createProject — POST /api/v1/newproject/create. Lightweight path
// that scaffolds a project (dir + git init + .gemba + beads DB +
// initial commit) with no plan tree, no Onboarder, no LLM.
// gm-root.17.13.
export async function createProject(
  body: CreateProjectRequest,
  opts: { nonce?: string } = {}
): Promise<RatifyResponse> {
  return apiFetch<RatifyResponse>('/v1/newproject/create', {
    method: 'POST',
    body: JSON.stringify(body),
    headers: {
      'Content-Type': 'application/json',
      [CONFIRM_HEADER]: opts.nonce ?? freshNonce(),
    },
  });
}

// probeOnboarder — GET /api/v1/onboarder/probe. Reports whether the
// Onboarder persona host can resolve an LLM client. The SPA reads
// this on board empty-state to decide whether to render the
// "Plan with the Onboarder →" CTA. gm-root.17.13.
export async function probeOnboarder(): Promise<OnboarderProbeResponse> {
  return apiFetch<OnboarderProbeResponse>('/v1/onboarder/probe', {
    method: 'GET',
  });
}

File: internal/server/newproject.go
// /api/v1/newproject/* — conversational project-creation flow
// (gm-root.17; design: docs/design/newproject.md).
//
// This file ships the server-side backing for the /new SPA route. The
// route runs a transient Onboarder-persona conversation against the
// `newproject` skill, then commits the resulting plan tree as a real
// project via an atomic transaction (gm-root.17.6).
//
// Wire contract: web/src/api/newproject.ts is the canonical shape. The
// Go types here mirror those typed-client interfaces verbatim — JSON
// tags use the design-doc PascalCase field names so the SPA does not
// need a separate translation layer.
//
// Three endpoints:
//   POST /api/v1/newproject/start          — mint a session
//   POST /api/v1/newproject/:id/turn       — submit an operator message
//   POST /api/v1/newproject/:id/ratify     — atomic commit
//
// The skill (gm-root.17.5) and the Onboarder persona host
// (gm-root.17.10) are out of scope for gm-root.17.6 — start/turn store
// state in-memory and emit deterministic stub replies. The skill plugs
// in via the SkillTurner interface (AttachNewProject), which the
// onboarder bead will implement against the real LLM client. Until
// then a stub turner ships so the SPA keeps working in real-mode.
//
// Atomic ratification (gm-root.17.6 — primary deliverable): the
// transaction lives in newproject_ratify.go. This file owns the
// HTTP plumbing, session state, and the in-memory store.

package server

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"errors"
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/go-chi/chi/v5"

	"github.com/GembaCore/gemba-core/internal/server/httperr"
	"github.com/GembaCore/gemba-core/internal/skills/newproject"
)

// ─── Wire types (re-exported from internal/skills/newproject) ───────
//
// The skill package owns the canonical schema (see
// internal/skills/newproject/types.go). This package re-exports the
// types under the same name so server-package code and tests don't
// have to reach across the import boundary every time, and so the
// SPA's contract (web/src/api/newproject.ts) has a single Go-side
// declaration to mirror.

// DraftBead is the leaf of the plan tree. See
// internal/skills/newproject for the field-level documentation; the
// JSON wire shape is canonical at that location.
type DraftBead = newproject.DraftBead

// DraftEpic is one epic in the plan tree.
type DraftEpic = newproject.DraftEpic

// DraftMilestone is one milestone in the plan tree.
type DraftMilestone = newproject.DraftMilestone

// ChangeRef points at the most-recent skill edit.
type ChangeRef = newproject.ChangeRef

// NewProjectState is the authoritative state for one /new session.
type NewProjectState = newproject.NewProjectState

// emptyNewProjectState returns the skill's canonical EmptyState.
// Keeps the existing call sites working without churn.
func emptyNewProjectState() NewProjectState {
	return newproject.EmptyState()
}

// ─── Wire envelopes ─────────────────────────────────────────────────

type startNewProjectResponse struct {
	SessionID string          `json:"session_id"`
	State     NewProjectState `json:"state"`
	Greeting  string          `json:"greeting"`
}

type turnRequest struct {
	Message string                 `json:"message"`
	Edits   map[string]interface{} `json:"edits,omitempty"`
}

type turnResponse struct {
	State   NewProjectState `json:"state"`
	Reply   string          `json:"reply"`
	ReplyID string          `json:"reply_id"`
	ReplyAt string          `json:"reply_at"`
}

type ratifyRequest struct {
	State NewProjectState `json:"state"`
}

// createRequest is the body for POST /api/v1/newproject/create — the
// lightweight project-creation path that does NOT spawn the Onboarder
// persona or call an LLM. The handler builds a minimal NewProjectState
// (empty Milestones[]) and runs the same atomic ratify transaction
// the conversational flow uses, so the operator gets a project dir,
// git init, .gemba/workspace.toml, and an empty beads DB without ever
// touching a chat client (gm-root.17.13).
type createRequest struct {
	ProjectName string `json:"project_name"`
	Description string `json:"description"`
}

// onboarderProbeResponse is the body of GET /api/v1/onboarder/probe.
// The SPA reads this on board empty-state to decide whether to render
// the "Plan with the Onboarder" CTA. When unavailable, the reason
// carries the canonical operator-facing diagnostic.
type onboarderProbeResponse struct {
	Available bool   `json:"available"`
	Reason    string `json:"reason,omitempty"`
}

// RatifyResponse is the success envelope POST /ratify returns. The
// SPA owns post-ratify navigation (gm-root.17.7); the server returns
// the new project's filesystem root, its display name, and the seeded
// milestone / epic counts so the handoff screen can render
// "seeded N milestones and M epics" without re-querying the new
// workspace's beads database.
type RatifyResponse struct {
	ProjectPath    string `json:"project_path"`
	ProjectName    string `json:"project_name"`
	MilestoneCount int    `json:"milestone_count"`
	EpicCount      int    `json:"epic_count"`
	// SeedWarnings carries any non-fatal failures from the FTUX seed
	// steps (.gemba/agents.toml, .gemba/personas/, CLAUDE.md — gm-root.24).
	// These are best-effort additions: a failure here does NOT roll back
	// the project — the SPA renders a "partial seed" toast so the
	// operator can hand-fix and proceed. Empty (or omitted) means every
	// seed step succeeded.
	SeedWarnings []string `json:"seed_warnings,omitempty"`
}

// ─── Session store ──────────────────────────────────────────────────

// NewProjectSession is the in-memory record for one /new conversation.
// Lives only in server process memory — refresh / restart / Ctrl-C
// discards it (one-shot persistence is by design; see
// docs/design/newproject.md §One-shot persistence).
type NewProjectSession struct {
	ID        string
	State     NewProjectState
	StartedAt time.Time
}

// NewProjectStore is the persistence interface for /new sessions.
// Mirrors the bootstrap / walk store pattern: callers hand pre-built
// records to mutation methods, the store concerns itself only with
// id + lifecycle.
type NewProjectStore interface {
	// Start allocates a new session with the empty state and returns
	// the persisted record so the handler can read the minted id back
	// without a second Get.
	Start(ctx context.Context) (NewProjectSession, error)

	// Get returns the session by id. ErrNewProjectNotFound when
	// unknown.
	Get(ctx context.Context, id string) (NewProjectSession, error)

	// Update replaces the state of an existing session. Returns
	// ErrNewProjectNotFound if the id is unknown.
	Update(ctx context.Context, id string, state NewProjectState) error

	// Delete drops the session from the store. Used after a
	// successful ratify to free memory; idempotent — deleting an
	// unknown id is not an error.
	Delete(ctx context.Context, id string)
}

// ErrNewProjectNotFound is returned by NewProjectStore when the
// session id is unknown.
var ErrNewProjectNotFound = errors.New("newproject: session not found")

// memoryNewProjectStore is the in-memory implementation. Capacity
// unbounded — sessions are operator-driven (a few per workspace ever)
// so an LRU is not worth the complexity in v1.
type memoryNewProjectStore struct {
	mu  sync.RWMutex
	now func() time.Time
	by  map[string]*NewProjectSession
}

// NewMemoryNewProjectStore returns a fresh in-memory store. cmd/gemba
// serve calls this once at boot and hands the result to
// AttachNewProject.
func NewMemoryNewProjectStore() NewProjectStore {
	return &memoryNewProjectStore{
		now: time.Now,
		by:  map[string]*NewProjectSession{},
	}
}

func (s *memoryNewProjectStore) Start(_ context.Context) (NewProjectSession, error) {
	id := newSessionID()
	rec := &NewProjectSession{
		ID:        id,
		State:     emptyNewProjectState(),
		StartedAt: s.now().UTC(),
	}
	s.mu.Lock()
	s.by[id] = rec
	s.mu.Unlock()
	return *rec, nil
}

func (s *memoryNewProjectStore) Get(_ context.Context, id string) (NewProjectSession, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	rec, ok := s.by[id]
	if !ok {
		return NewProjectSession{}, ErrNewProjectNotFound
	}
	return *rec, nil
}

func (s *memoryNewProjectStore) Update(_ context.Context, id string, state NewProjectState) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	rec, ok := s.by[id]
	if !ok {
		return ErrNewProjectNotFound
	}
	rec.State = state
	return nil
}

func (s *memoryNewProjectStore) Delete(_ context.Context, id string) {
	s.mu.Lock()
	delete(s.by, id)
	s.mu.Unlock()
}

// newSessionID mints a 16-byte hex session id with a "newproj-"
// prefix for log-grep convenience.
func newSessionID() string {
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		return "newproj-" + time.Now().UTC().Format("20060102T150405.000000000")
	}
	return "newproj-" + hex.EncodeToString(b)
}

// ─── Skill turner (pluggable LLM contact point) ─────────────────────

// SkillTurner is the contract the Onboarder persona host (gm-root.17.10)
// implements: take a state + an operator message + any in-place edits
// and return the next state + a reply for the conversation pane.
//
// Until the real persona host lands, a stub implementation ships so
// the SPA keeps working in real mode. The stub mirrors the e2e fixture's
// deterministic plan-tree mutation so screenshot tests stay stable.
type SkillTurner interface {
	Turn(ctx context.Context, in NewProjectState, message string, edits map[string]interface{}) (out NewProjectState, reply string, err error)
	// Greeting returns the assistant's opening line on /start. Empty
	// is a valid greeting (the SPA renders nothing).
	Greeting() string
	// Probe is invoked on /start before allocating a session and
	// reports whether the turner can serve the request. Real-mode
	// implementations return an error wrapping the operator-facing
	// diagnostic (e.g. "no LLM client configured") so the handler
	// can return 503 with the diagnostic in the body. Stub turners
	// return nil. May be called concurrently.
	Probe(ctx context.Context) error
}

// stubSkillTurner is a deterministic placeholder used until the
// real Onboarder persona ships (gm-root.17.10). It mirrors the e2e
// fixture's behaviour: the first turn seeds a single placeholder
// milestone; subsequent turns append epics under it.
type stubSkillTurner struct{}

// NewStubSkillTurner returns the placeholder turner. Production
// builds use the real Onboarder when AttachNewProject(... real-turner)
// is wired (gm-root.17.10).
func NewStubSkillTurner() SkillTurner { return stubSkillTurner{} }

func (stubSkillTurner) Probe(_ context.Context) error { return nil }

func (stubSkillTurner) Greeting() string {
	return "Hi — what are we building? Tell me the rough shape (web app, library, ops tool, …) and I'll propose milestones, epics, and beads."
}

func (stubSkillTurner) Turn(_ context.Context, in NewProjectState, message string, _ map[string]interface{}) (NewProjectState, string, error) {
	out := in
	out.Turn = in.Turn + 1
	msg := strings.TrimSpace(message)
	if out.ProjectName == "" && msg != "" {
		// First turn seeds a project name from the user's message
		// (truncated to a reasonable length). Real skill replaces this
		// with a parsed intent.
		name := msg
		if len(name) > 48 {
			name = name[:48]
		}
		out.ProjectName = name
		out.Description = msg
	}
	if len(out.Milestones) == 0 {
		out.Milestones = []DraftMilestone{{
			Title:       "M1 — define scope",
			Description: "Initial milestone seeded by the stub Onboarder.",
			Labels:      []string{},
			Skills:      []string{},
			Epics:       []DraftEpic{},
		}}
		out.LastChange = ChangeRef{Path: "milestone:0", Kind: "added", Summary: "Drafted milestone 1."}
		return out, "Drafted 1 milestone. Tell me more and I'll add epics + beads.", nil
	}
	// Subsequent turns append a placeholder epic under the first
	// milestone so the SPA's diff badge has something to render.
	epicIdx := len(out.Milestones[0].Epics)
	out.Milestones[0].Epics = append(out.Milestones[0].Epics, DraftEpic{
		Title:       "E" + newProjectItoa(epicIdx+1) + " — placeholder epic",
		Description: "Stub epic; the real Onboarder will replace this.",
		Labels:      []string{},
		Skills:      []string{},
		Beads:       []DraftBead{},
	})
	out.LastChange = ChangeRef{
		Path:    "milestone:0/epic:" + newProjectItoa(epicIdx),
		Kind:    "added",
		Summary: "Added placeholder epic under milestone 1.",
	}
	return out, "Added a placeholder epic. Edit it in the preview pane or keep typing.", nil
}

// newProjectItoa is a tiny helper for the stub turner — renamed to avoid
// colliding with the package-level itoa in consults_apply.go.
func newProjectItoa(i int) string {
	if i == 0 {
		return "0"
	}
	neg := false
	if i < 0 {
		neg = true
		i = -i
	}
	var buf [20]byte
	pos := len(buf)
	for i > 0 {
		pos--
		buf[pos] = byte('0' + i%10)
		i /= 10
	}
	if neg {
		pos--
		buf[pos] = '-'
	}
	return string(buf[pos:])
}

// ─── Router glue ────────────────────────────────────────────────────

// AttachNewProject binds the /new session store + skill turner +
// ratify executor (gm-root.17.6). Calling with a nil store removes
// the binding (handlers return 503). cmd/gemba serve calls this once
// at boot; tests inject fixtures per case.
//
// turner may be nil — when so, NewStubSkillTurner takes over. ratifier
// may be nil too — when so, /ratify returns 503 (the conversation
// surface still works for SPA preview / screenshot purposes).
func (r *Router) AttachNewProject(store NewProjectStore, turner SkillTurner, ratifier NewProjectRatifier) {
	r.newProjectStore = store
	r.newProjectTurner = turner
	r.newProjectRatifier = ratifier
}

func newProjectAdaptorUnavailable(w http.ResponseWriter) {
	httperr.Write(w, http.StatusServiceUnavailable,
		"adaptor_not_configured", "newproject store not registered")
}

// ─── Handlers ───────────────────────────────────────────────────────

func (r *Router) newProjectStart(w http.ResponseWriter, req *http.Request) {
	if r.newProjectStore == nil {
		newProjectAdaptorUnavailable(w)
		return
	}
	turner := r.newProjectTurner
	if turner == nil {
		turner = NewStubSkillTurner()
	}
	// Spawn-failure path (gm-root.17.10): real-mode turners (the
	// Onboarder persona) return an error here when they can't
	// resolve a chat client. Surface as 503 with the operator-facing
	// diagnostic so the SPA's /new route can render it verbatim
	// without parsing a 500 stack trace.
	if err := turner.Probe(req.Context()); err != nil {
		httperr.Write(w, http.StatusServiceUnavailable,
			"no_llm_client", err.Error())
		return
	}
	rec, err := r.newProjectStore.Start(req.Context())
	if err != nil {
		httperr.WriteError(w, err)
		return
	}
	writeJSON(w, http.StatusCreated, startNewProjectResponse{
		SessionID: rec.ID,
		State:     rec.State,
		Greeting:  turner.Greeting(),
	})
}

func (r *Router) newProjectTurn(w http.ResponseWriter, req *http.Request) {
	if r.newProjectStore == nil {
		newProjectAdaptorUnavailable(w)
		return
	}
	turner := r.newProjectTurner
	if turner == nil {
		turner = NewStubSkillTurner()
	}
	id := strings.TrimSpace(chi.URLParam(req, "id"))
	if id == "" {
		httperr.Write(w, http.StatusBadRequest, "bad_request",
			"session id is required")
		return
	}
	var body turnRequest
	if err := decodeJSONBody(req, &body); err != nil {
		httperr.Write(w, http.StatusBadRequest, "bad_request",
			"invalid JSON body: "+err.Error())
		return
	}
	sess, err := r.newProjectStore.Get(req.Context(), id)
	if err != nil {
		writeNewProjectError(w, err)
		return
	}
	out, reply, err := turner.Turn(req.Context(), sess.State, body.Message, body.Edits)
	if err != nil {
		httperr.WriteError(w, err)
		return
	}
	if err := r.newProjectStore.Update(req.Context(), id, out); err != nil {
		writeNewProjectError(w, err)
		return
	}
	now := time.Now().UTC()
	writeJSON(w, http.StatusOK, turnResponse{
		State:   out,
		Reply:   reply,
		ReplyID: "reply-" + newSessionID(),
		ReplyAt: now.Format(time.RFC3339Nano),
	})
}

func (r *Router) newProjectRatify(w http.ResponseWriter, req *http.Request) {
	if r.newProjectStore == nil {
		newProjectAdaptorUnavailable(w)
		return
	}
	if r.newProjectRatifier == nil {
		httperr.Write(w, http.StatusServiceUnavailable,
			"adaptor_not_configured",
			"newproject ratifier not registered")
		return
	}
	id := strings.TrimSpace(chi.URLParam(req, "id"))
	if id == "" {
		httperr.Write(w, http.StatusBadRequest, "bad_request",
			"session id is required")
		return
	}
	var body ratifyRequest
	if err := decodeJSONBody(req, &body); err != nil {
		httperr.Write(w, http.StatusBadRequest, "bad_request",
			"invalid JSON body: "+err.Error())
		return
	}
	sess, err := r.newProjectStore.Get(req.Context(), id)
	if err != nil {
		writeNewProjectError(w, err)
		return
	}
	// Stale-modal guard: the SPA POSTs the snapshot the operator
	// confirmed in the modal. Reject if the live session state has
	// drifted (turn count differs) so a background turn racing the
	// confirm click cannot commit a tree the operator did not see.
	if body.State.Turn != sess.State.Turn {
		httperr.Write(w, http.StatusConflict, "stale_state",
			"session state changed since the ratify modal opened; reopen and retry")
		return
	}
	resp, err := r.newProjectRatifier.Ratify(req.Context(), body.State)
	if err != nil {
		writeRatifyError(w, err)
		return
	}
	// Successful ratify — drop the session so a stale id can't double-
	// commit on a delayed retry that bypassed the nonce cache (e.g.
	// the operator hits Back and clicks Ratify again).
	r.newProjectStore.Delete(req.Context(), id)
	writeJSON(w, http.StatusOK, resp)
}

// writeNewProjectError maps NewProjectStore errors onto the shared
// envelope. ErrNewProjectNotFound → 404; everything else flows
// through the generic mapper.
func writeNewProjectError(w http.ResponseWriter, err error) {
	switch {
	case errors.Is(err, ErrNewProjectNotFound):
		httperr.Write(w, http.StatusNotFound, "newproject_not_found", err.Error())
	default:
		httperr.WriteError(w, err)
	}
}

// newProjectCreate is the lightweight project-creation handler. It
// runs the same atomic ratify transaction the conversational flow
// uses, but with an empty Milestones[] and ProjectName/Description
// straight from the request — no session, no Onboarder, no LLM
// (gm-root.17.13). Operators who want to scaffold a project and
// manage beads manually never touch a chat client.
func (r *Router) newProjectCreate(w http.ResponseWriter, req *http.Request) {
	if r.newProjectRatifier == nil {
		httperr.Write(w, http.StatusServiceUnavailable,
			"adaptor_not_configured",
			"newproject ratifier not registered")
		return
	}
	var body createRequest
	if err := decodeJSONBody(req, &body); err != nil {
		httperr.Write(w, http.StatusBadRequest, "bad_request",
			"invalid JSON body: "+err.Error())
		return
	}
	name := strings.TrimSpace(body.ProjectName)
	if name == "" {
		httperr.Write(w, http.StatusBadRequest, "bad_request",
			"project_name is required")
		return
	}
	state := emptyNewProjectState()
	state.ProjectName = name
	state.Description = strings.TrimSpace(body.Description)
	resp, err := r.newProjectRatifier.Ratify(req.Context(), state)
	if err != nil {
		writeRatifyError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, resp)
}

// onboarderProbe reports whether the Onboarder persona host can
// resolve an LLM client. Returns 200 in both available + unavailable
// cases — the body's `available` flag is the operative signal. The
// SPA uses this to gate the optional "Plan with the Onboarder" CTA
// on the board's empty-state without invoking the (heavier)
// /api/v1/newproject/start probe (gm-root.17.13).
func (r *Router) onboarderProbe(w http.ResponseWriter, req *http.Request) {
	turner := r.newProjectTurner
	if turner == nil {
		writeJSON(w, http.StatusOK, onboarderProbeResponse{Available: true})
		return
	}
	if err := turner.Probe(req.Context()); err != nil {
		writeJSON(w, http.StatusOK, onboarderProbeResponse{
			Available: false,
			Reason:    err.Error(),
		})
		return
	}
	writeJSON(w, http.StatusOK, onboarderProbeResponse{Available: true})
}
